examples/cpp/waku.cpp

```diff
@@ -0,0 +1,336 @@
+    fcntl(0, F_SETFL, fcntl(0, F_GETFL) | O_NONBLOCK);
+}
+
+void set_scanf_to_block() {
+    fcntl(0, F_SETFL, fcntl(0, F_GETFL) ^ O_NONBLOCK);
+}
+
+void handle_user_input() {
+    char cmd[1024];
+    memset(cmd, 0, 1024);
+    int numRead = read(0, cmd, 1024);
+    if (numRead <= 0) {
+        return;
+    }
+
+    int c;
+    while ( (c = getchar()) != '\n' && c != EOF ) { }
+
+    switch (atoi(cmd))
+    {
+    case SUBSCRIBE_TOPIC_MENU:
+    {
+        printf("Indicate the Pubsubtopic to subscribe:\n");
+        set_scanf_to_block();
+        char pubsubTopic[128];
+        scanf("%127s", pubsubTopic);
+        // if (!waku_relay_subscribe(pubsubTopic, &mResp)) {
+        //     printf("Error subscribing to PubsubTopic: %s\n", mResp->data);
+        // }
+        // printf("Waku Relay subscription response: %s\n", mResp->data);
+
+        set_scanf_to_not_block();
+        show_main_menu();
+    }
+    break;
+
+    case CONNECT_TO_OTHER_NODE_MENU:
+        printf("Connecting to a node. Please indicate the peer Multiaddress:\n");
+        printf("e.g.: /ip4/127.0.0.1/tcp/60001/p2p/16Uiu2HAmVFXtAfSj4EiR7mL2KvL4EE2wztuQgUSBoj2Jx2KeXFLN\n");
+        set_scanf_to_block();
+        char peerAddr[512];
+        scanf("%511s", peerAddr);
+        // if (!waku_connect(peerAddr, 10000 /* timeoutMs */, &mResp)) {
+        //     printf("Couldn't connect to the remote peer: %s\n", mResp->data);
+        // }
+        set_scanf_to_not_block();
+        show_main_menu();
+    break;
+
+    case PUBLISH_MESSAGE_MENU:
+    {
+        set_scanf_to_block();
+        printf("Indicate the Pubsubtopic:\n");
+        char pubsubTopic[128];
+        scanf("%127s", pubsubTopic);
+
+        printf("Type the message tp publish:\n");
+        char msg[1024];
+        scanf("%1023s", msg);
+
+        char jsonWakuMsg[1024];
+        std::vector<char> msgPayload;
+        b64_encode(msg, strlen(msg), msgPayload);
+
+        // waku_content_topic("appName",
+        //                     1,
+        //                     "contentTopicName",
+        //                     "encoding",
+        //                     &mResp);
+
+        // snprintf(jsonWakuMsg,
+        //          1024,
+        //          "{\"payload\":\"%s\",\"content_topic\":\"%s\"}",
+        //          msgPayload, mResp->data);
+
+        // free(msgPayload);
+
+        // waku_relay_publish(pubsubTopic, jsonWakuMsg, 10000 /*timeout ms*/, &mResp);
+        // printf("waku relay response [%s]\n", mResp->data);
+        // set_scanf_to_not_block();
+        show_main_menu();
+    }
+    break;
+
+    case MAIN_MENU:
+        break;
+    }
+}
+
+// End of UI program logic
+
+void show_help_and_exit() {
+    printf("Wrong parameters\n");
+    exit(1);
+}
+
+void event_handler(const char* msg, size_t len) {
+    printf("Receiving message %s\n", msg);
+}
+
+void handle_error(const char* msg, size_t len) {
+    printf("Error: %s\n", msg);
+    exit(1);
+}
+
+template <class F>
+auto cify(F&& f) {
```

Comment:
I'm not pretty sure that wrapping around lambdas is necessary. Maybe enough to simple give lamba in place to the call of C function.

Reply:
Good point! Afaik, this is needed. We'll revisit that point in the future.